### ISSUE TITLE
Mismatch in interface between request and tough-cookie for functions getCookieString and setCookie.

### ISSUE THREAD

**Issue by Eegee** [request/request#1720]
There is a mismatch in the interface between request and tough-cookie, regarding function getCookieString (and indirectly getCookies).

When I want to assign a tough-cookie jar, request does the following in "Request.prototype.jar = function (jar)" :

cookies = targetCookieJar.getCookieString(urihref)

However, tough-cookie requires a callback as the second or third parameter in CookieJar.prototype.getCookieString, it is not an optional parameter. See https://github.com/SalesforceEng/tough-cookie#getcookiescurrenturl-options-cberrcookies and https://github.com/SalesforceEng/tough-cookie#getcookiestring

Because of this, in getCookieString, tough-cookie gets a uri as the first and only parameter. That goes wrong and breaks, because it pops the last argument which it believes is the callback function but really is the uri. It replaces that with a new callback function that returns a cookie string (thinking it replaces the original callback argument but really replaces the uri argument). It applies that to getCookies, so that function only gets the new callback as the first parameter and not the uri.

For now I've updated request.js in my local project to provide a valid callback function to getCookieString. This callback function needs to set the cookie HTTP-header, because getCookieString does not return the cookie string, it must go through the callback. So I've moved that piece of code and variable 'cookies' in function Request.prototype.jar is no longer used:

```js
    if (targetCookieJar) {
      targetCookieJar.getCookieString(urihref, function (err, cookies) {
        //if need cookie and cookie is not empty
        if (!err && cookies && cookies.length) {
            if (self.originalCookieHeader) {
                // Don't overwrite existing Cookie header
                self.setHeader('cookie', self.originalCookieHeader + '; ' + cookies)
            } else {
                self.setHeader('cookie', cookies)
            }
        }
      });
    }
```

So could you please take a look at getCookiestring and its interface? Thank you.

**Comment by Eegee** [comment]
There's the same problem for function setCookie. The callback parameter is mandatory but is not passed:
targetCookieJar.setCookie(cookie, self.uri.href, {ignoreError: true})

See https://www.npmjs.com/package/tough-cookie#setcookie-cookieorstring-currenturl-options-cb-err-cookie

**Comment by simov** [comment]
@lalitkapoor want to help? :)

**Comment by lalitkapoor** [comment]
@Eegee In request the cookie implementation is using the synchronous methods of tough-cookie. If you look at https://github.com/request/request/blob/master/lib/cookies.js you'll see that they make all sync calls to the tough-cookie api. 

I believe this was the legacy way of doing things, before async store support was added to tough-cookie; and for backwards compatibility we haven't changed the implementation to be async in request, hence the mismatch in the function names now.

That said, the cookie implementation should be working correctly in accordance with the docs in request. Please take a look at the tests here to see more examples - https://github.com/request/request/blob/master/tests/test-cookies.js

Please let me know if you have any more questions. You shouldn't need to patch request to get it working.

**Comment by Eegee** [comment]
Okay, I feel stupid now.... :D

What I did wrong was using a separate tough-cookie instance, created with new CookieJar() instead of using request's function to create one: new request.jar().
That's why I saw it going through async methods instead of the synchronous ones.

Oh well, learning Node.JS as we go... to quote a certain A. Einstein: “Anyone who has never made a mistake has never tried anything new.”
